File: Java/Moderate/Arrays and Strings/MissingRanges.java
class Solution {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> result = new ArrayList<String>();
        if(nums.length==0)
        {
        	if(lower==upper)
        	{
        		result.add(""+lower);
        	}
        	else
        	{
        		int firstNumber = lower;
        		int secondNumber = upper;
        		result.add(""+firstNumber+"->"+secondNumber);
        	}
        	return result;
        }
        boolean lowerFound = false;
        for(int i=0;i<nums.length;i++)
        {
        	if(nums[i]>=lower && !lowerFound)
        	{
        		lowerFound = true;;
        		if(nums[i]!=lower)
        		{
        			result.add(addFirsttoList(lower,nums[i]));
        		}
        	}
        	else if(nums[i]>=upper && lowerFound)
        	{
        		if(nums[i]==upper && upper-nums[i-1]>1)
        		{
        			result.add(addLasttoList(nums[i-1],upper-1));
        		}
        		break;
        	}
        	else if(nums[i]-nums[i-1]>1 && lowerFound)
        	{
        		result.add(addOtherstoList(nums[i-1],nums[i]));
        	}
        }
        if(nums[nums.length-1]<upper)
        {
        	if(lowerFound)
        	{
        		result.add(addLasttoList(nums[nums.length-1],upper));
        	}
        	else
        	{
        		if(lower==upper)
	        	{
	        		result.add(""+lower);
	        	}
	        	else
	        	{
	        		int firstNumber = lower;
	        		int secondNumber = upper;
	        		result.add(""+firstNumber+"->"+secondNumber);
	        	}
        	}
        }
        return result;
    }
    private String addFirsttoList(int smaller,int greater)
    {
    	if(greater-smaller==1)
    	{
    		return smaller+"";
    	}
    	else
    	{
    		int first = smaller;
    		int second = greater-1;
    		return ""+first+"->"+second;
    	}
    }
    private String addOtherstoList(int smaller,int greater)
    {
    	if(greater-smaller==2)
    	{
    		int first = smaller+1;
    		return ""+first;
    	}
    	else
    	{
    		int first = smaller+1;
    		int second = greater-1;
    		return first+"->"+second;
    	}
    }
    private String addLasttoList(int smaller,int greater)
    {
    	if(greater-smaller==1)
    	{
    		return ""+greater;
    	}
    	else
    	{
    		int first = smaller+1;
    		int second = greater;
    		return first+"->"+second;
    	}
    }
}

File: CSharp/Medium/Array and Strings/MissingRanges.cs
/*
Example:

Input: nums = [0, 1, 3, 50, 75], lower = 0 and upper = 99,
Output: ["2", "4->49", "51->74", "76->99"]
*/
public class Solution {
    public IList<string> FindMissingRanges(int[] nums, int lower, int upper) {
 		IList<string> result = new List<string>();
 		int lastFoundValue = -1;
 		bool lowerFound = false;
 		for(int i=0;i<nums.Length;i++)
 		{
 			if(!lowerFound && nums[i]>=lower)
 			{
 				if(nums[i]!=lower)
	 			{
	 				result.Add(GetFormatedString(lower,nums[i]-1));
	 			}
 				lastFoundValue = nums[i];
				lowerFound = true;
 			}
 			else if(lowerFound && nums[i] >= lastFoundValue+1 && nums[i]>=upper)
 			{
 				if(nums[i]>=lastFoundValue+1 && nums[i]>upper)
 				{
 					result.Add(GetFormatedString(lastFoundValue+1,upper));
 				}
 				else if(nums[i]>lastFoundValue+1 && nums[i]==upper)
 				{
 					result.Add(GetFormatedString(lastFoundValue+1,upper-1));	
 				}
 				lastFoundValue = upper;
				break;
 			}
 			else if(lowerFound && nums[i] >= lastFoundValue+1)
 			{
 				if(nums[i] != lastFoundValue+1)
 				{
 					result.Add(GetFormatedString(lastFoundValue+1,nums[i]-1));
 				}
 				lastFoundValue = nums[i];
 			}
 		}  
 		if(lowerFound && lastFoundValue<upper)
 		{
 			result.Add(GetFormatedString(lastFoundValue+1,upper));
 		}    
 		if(!lowerFound)
 		{
 			result.Add(GetFormatedString(lower,upper));
 		}
 		return result; 
    }

    private string GetFormatedString(int num1, int num2)
    {
    	if(num1==num2)
    	{
    		return num1.ToString();
    	}
    	else
    	{
    		StringBuilder sb = new StringBuilder();
    		sb.Append(num1);
    		sb.Append("->");
    		sb.Append(num2);
    		return sb.ToString();
    	}
    }
}